### DISCUSSION TITLE
When using rector in conjunction with ecs, how to be sure declare(strict_types=1) is the first statement?

### DISCUSSION THREAD

**Discussion by alexandreelise** [rectorphp/rector#7414] [Q&A] [answered]
Hi Rector Team,
I now grasp more how rector and ecs work. I love the unified config file structure that works generically for both projects and many more.

Now my question is how to prevent rector and ecs stepping on their toes each other?

When I run ecs without the strict rule but rector with code quality it puts the use statements at the top of the file before the declare(strict_types=1); statement which should be the first statement right after <?php when present.

Is there a way to prevent this behavior? @TomasVotruba @samsonasik ?

**Comment by TomasVotruba** [comment, marked as answer]
Hi, that should not happen.
we'll need a demo link with particular Rector rule and invalid code.

**Reply by alexandreelise** [reply]
Hi @TomasVotruba,

I struggled like a worm to make a failing test to reproduce the supposed "bug" but couldn't reproduce it locally even if a confined reduced set of classes. Rector and ECS beat me flawlessly with a perfect score. I saw the message in my mind "You lose! PERFECT!" "Game Over" "Insert a coin".

Anyway, joke aside, this experience a creating a reduce set repo helped me grasp the best pratices in our industry and more importantly understand more and more your everyday work answering to random developers questions around the world. I am so happy to learn rector and you seem to be such a nice guy not like many developer around there. Thanks for your help you and your awesome team and for such a cool futuristic tool called rector. I will make a vow to master this tool to it's limits and try to contribute back as much as I can If I am not behind an NDA.

Would you like me to share my local repo anyway as an example repo?

Have a great day.

**Reply by TomasVotruba** [reply]
Hi @alexandreelise ,

thanks for sharing your story :+1: Much appreciated. First steps are the hardest, and that's how learn :pray: 

Feel free to send us an NDA and share your repository. We might be able to help to figure it out together in private:
https://getrector.org/contact

**Reply by alexandreelise** [reply]
NDA is not necessarily coming from me. Most of my projects I do are private projects for clients that don't want to share their work for obvious competitive advantange reasons. For this particular demo rector sample I feel that I can share it and it might help rector community. When I finally took the time to read the docs completely I saw the potential of FileProcessors. I 'll play around with them this weekend I wanna dig deeper behind the mirror.

**Reply by alexandreelise** [reply]
Here is the demo sample repo https://github.com/alexandreelise/rector-demo-sample.git

**Reply by TomasVotruba** [reply]
Thanks 👍

Could you add a GitHub Action with reproducible steps? That way we'll see what is wrong right in the CI and also have verification when its fixed.

**Reply by alexandreelise** [reply]
To be clear nothing is wrong anymore. I think I made a mistake on my end. I cannot reproduce the "bug".   locally. This repo is just an example for me and others nothing to do with the bug I told you about above. At first I tried to do it for the bug. But finally I decided to let it go to the github universe flying in alternative world between the stars and the code. BTW thanks for sharing this month jetbrains blog issue. Cool for rector and I found many new tools listed in the blog post useful and cool. Thanks.

**Reply by TomasVotruba** [reply]
Alright then 😌👍
Thanks for explaining

**Reply by alexandreelise** [reply]
No problem Tomas. I am happy to learn from a humble superstar like you.

**Reply by TomasVotruba** [reply]
Same for me, I'm learning from all the open-source leaders and every contributor like you, 
who can show me what else can be done better.

Thank you for warm words :bow:
